tools/state-parts-dump-check/src/cli.rs

```diff
@@ -555,7 +565,106 @@ async fn run_single_check(
 
     let duration = start.elapsed();
     tracing::info!("Time elapsed in downloading and validating the parts is: {:?}", duration);
-    Ok(StatePartsDumpCheckStatus::Done { epoch_height: epoch_height })
+    Ok(true)
+}
+
+// download and validate state headers for a single epoch and shard
+async fn check_headers(
+    chain_id: &String,
+    epoch_id: &EpochId,
+    epoch_height: u64,
+    shard_id: ShardId,
+    external: &ExternalConnection,
+) -> anyhow::Result<bool> {
+    let directory_path = external_storage_location_directory(
+        &chain_id,
+        &epoch_id,
+        epoch_height,
+        shard_id,
+        &StateFileType::StateHeader,
+    );
+    tracing::info!(directory_path, "the storage location for the state header being checked:");
+    if !external
+        .is_state_sync_header_stored_for_epoch(shard_id, chain_id, epoch_id, epoch_height)
+        .await?
+    {
+        tracing::info!(epoch_height, shard_id, "Waiting for header to be dumped.");
+        return Ok(false);
+    }
+
+    crate::metrics::STATE_SYNC_DUMP_CHECK_NUM_HEADERS_DUMPED
+        .with_label_values(&[&shard_id.to_string(), &chain_id.to_string()])
+        .set(1 as i64);
+
+    tracing::info!(shard_id, epoch_height, "Download and validate state header.");
+
+    let start = Instant::now();
+    let chain_id = chain_id.clone();
+    let epoch_id = epoch_id.clone();
+    let external = external.clone();
+
+    process_header_with_3_retries(chain_id, epoch_id, epoch_height, shard_id, external)
+        .await
+        .unwrap();
+
+    let duration = start.elapsed();
+    tracing::info!("Time elapsed in downloading and validating the header is: {:?}", duration);
+    Ok(true)
+}
+
+async fn run_single_check(
+    status: Option<StatePartsDumpCheckStatus>,
+    chain_id: String,
+    epoch_id: EpochId,
+    epoch_height: u64,
+    shard_id: ShardId,
+    state_root: StateRoot,
+    root_dir: Option<PathBuf>,
+    s3_bucket: Option<String>,
+    s3_region: Option<String>,
+    gcs_bucket: Option<String>,
+) -> anyhow::Result<StatePartsDumpCheckStatus> {
+    tracing::info!(
+        epoch_height,
+        %state_root,
+        "run_single_check for"
+    );
+    crate::metrics::STATE_SYNC_DUMP_CHECK_EPOCH_HEIGHT
+        .with_label_values(&[&shard_id.to_string(), &chain_id.to_string()])
+        .set(epoch_height as i64);
+
+    crate::metrics::STATE_SYNC_DUMP_CHECK_PROCESS_IS_UP
+        .with_label_values(&[&shard_id.to_string(), &chain_id.to_string()])
+        .set(1);
+
+    let external = create_external_connection(
+        root_dir.clone(),
+        s3_bucket.clone(),
+        s3_region.clone(),
+        gcs_bucket.clone(),
+    );
+
+    let (mut parts_done, mut headers_done) = match status {
+        Some(StatePartsDumpCheckStatus::Done { .. }) => (true, true),
```

Comment:
This logic plus the below `parts_done = parts_done || check_parts()` seems wrong to me. if the status is `StatePartsDumpCheckStatus::Done`, and the epoch has not advanced, then we continue in the loop in `run_loop_all_shards()` and don't call this function. Otherwise, this `StatePartsDumpCheckStatus::Done` status refers to the previous epoch right? In which case we don't want to `||` the old status with the new, but we want to start over from `StatePartsDumpCheckStatus::Waiting`. But that is only set at the very beginning when the `last_check_status` map doesn't have anything in it. So looks like this PR would make us report every epoch as `Done` as soon as the first one is done

Reply:
Good catch. To fix this, we can check if the epoch in the status is the same with the current epoch.